drivers/i2c/gpio_i2c_switch.c

```diff
@@ -0,0 +1,105 @@
+/*
+ * Copyright (c) 2023 Ayush Singh <ayushdevel1325@gmail.com>
+ * Copyright (c) 2021 Jason Kridner, BeagleBoard.org Foundation
+ * Copyright (c) 2020 Innoseis BV
+ *
+ * Based on i2c_tca9656a.c
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT gpio_i2c_switch
+
+#define GPIO_I2C_TRANSFER_DELAY_US 100
+
+#include <zephyr/kernel.h>
+#include <zephyr/device.h>
+#include <zephyr/devicetree.h>
+#include <zephyr/drivers/i2c.h>
+#include <zephyr/drivers/gpio.h>
+#include <zephyr/logging/log.h>
+
+LOG_MODULE_REGISTER(gpio_i2c_switch, CONFIG_I2C_LOG_LEVEL);
+
+struct gpio_i2c_switch_config {
+	const struct device *bus;
+	const struct gpio_dt_spec gpio;
+};
+
+struct gpio_i2c_switch_data {
+	struct k_mutex lock;
+};
+
+static int gpio_i2c_switch_configure(const struct device *dev, uint32_t dev_config)
+{
+	const struct gpio_i2c_switch_config *config = dev->config;
+
+	return i2c_configure(config->bus, dev_config);
+}
+
+static int gpio_i2c_switch_enable(const struct device *dev)
+{
+	const struct gpio_i2c_switch_config *config = dev->config;
+
+	return gpio_pin_configure(config->gpio.port, config->gpio.pin, GPIO_OUTPUT_HIGH);
+}
+
+static int gpio_i2c_switch_disable(const struct device *dev)
+{
+	const struct gpio_i2c_switch_config *config = dev->config;
+
+	return gpio_pin_configure(config->gpio.port, config->gpio.pin, GPIO_OUTPUT_LOW);
+}
+
+static int gpio_i2c_switch_transfer(const struct device *dev, struct i2c_msg *msgs,
+				    uint8_t num_msgs, uint16_t addr)
+{
+	int res;
+	struct gpio_i2c_switch_data *data = dev->data;
+	const struct gpio_i2c_switch_config *config = dev->config;
+
+	res = k_mutex_lock(&data->lock, K_MSEC(5000));
+	if (res != 0) {
+		return res;
+	}
+
+	gpio_i2c_switch_enable(dev);
+	k_busy_wait(GPIO_I2C_TRANSFER_DELAY_US);
+	res = i2c_transfer(config->bus, msgs, num_msgs, addr);
+	if (res != 0) {
+		goto disable_switch;
+	}
+	k_busy_wait(GPIO_I2C_TRANSFER_DELAY_US);
```

Comment:
hey thinking about the logic, is this wait needed? I understand the one before the transaction if the switch has any settling time, but then why would you wait BEFORE setting it back? This should be placed after `gpio_i2c_switch_disable`.

Reply:
Yes, it can be removed. I have also reduced the gpio toggle delay to 1 us after checking the [data sheet](https://www.ti.com/lit/ds/symlink/ts5a2066.pdf?ts=1699344464626&ref_url=https%253A%252F%252Fwww.google.com%252F) according to which the max on/off time is 12.9 ns.

Reply:
I believe there was a delay in case there was any propagation time through the controller. The intent was to avoid cutting the bus off before the transaction is completed at the hardware. As long as we are past the ACK, all should be good and no wait after the transaction *before* flipping the I/O is needed. And, as it seems, the delay time from flipping the I/O and having it propagate to the switch before new I2C transfers can happen seems negligable. Thanks for going through all this effort to clean up the hacks and sorry it slipped through the initial upstream submission.

Reply:
Of cousre, it's a good addition, sure there's more board around that could need this. Also now that I look at it again I realized I missed a misuse of the GPIO API :-) let me put together a quick followup for that.